Add invariant to VecDeque::pop_* that len < cap if pop successful

Philippe-Cholet opened the pull request:
Similar to #114370 for VecDeque instead of Vec.

I initially come from https://github.com/rust-itertools/itertools/pull/899 where we noticed that `pop_front;push_back;` was slower than expected so @scottmcm suggested I file an issue which lead to https://internals.rust-lang.org/t/vecdeque-pop-front-push-back/20483 where **kornel** mentionned #114334 (fixed by #114370).

This is my first time with codegen tests, I based the test on what was done for Vec.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8434625557/job/23098228461) [(plain)](https://github.com/rust-lang/rust/commit/9bc4f28b883b6293232b19ad55bdd47e93223f4d/checks/23098228461/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.7s done
#16 DONE 32.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling unwind v0.0.0 (/checkout/library/unwind)
error[E0133]: call to unsafe function `assume` is unsafe and requires unsafe block
    --> library/alloc/src/collections/vec_deque/mod.rs:1617:13
     |
1617 |             core::intrinsics::assume(self.len < self.capacity());
     |
     = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: call to unsafe function `assume` is unsafe and requires unsafe block
error[E0133]: call to unsafe function `assume` is unsafe and requires unsafe block
    --> library/alloc/src/collections/vec_deque/mod.rs:1642:13
     |
1642 |             core::intrinsics::assume(self.len < self.capacity());
     |
     = note: consult the function's documentation for information on how to avoid undefined behavior

For more information about this error, try `rustc --explain E0133`.
```

</details>

Philippe-Cholet commented:
I just saw that the `Vec::pop` optimization was updated to `core::hint::assert_unchecked(self.len < self.capacity());` so I probably should too.

Nilstrieb commented:
Yeah, that's nicer.
Gonna check compile time performance in the meantime (i don't expect many results), don't push the change until the try build is done (~2h) or it will be cancelled
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit e822937428e80bb9609a40b298d1c8b4e157c6ef with merge debc5abdd3fb55de560daf90c834438ed3ebbc31...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e822937428e80bb9609a40b298d1c8b4e157c6ef","merge_sha":"debc5abdd3fb55de560daf90c834438ed3ebbc31"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8437265441/job/23111805830)
Build commit: debc5abdd3fb55de560daf90c834438ed3ebbc31 (`debc5abdd3fb55de560daf90c834438ed3ebbc31`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8437265441/job/23111805830"},"merge_sha":"debc5abdd3fb55de560daf90c834438ed3ebbc31"} -->

rust-timer commented:
Queued debc5abdd3fb55de560daf90c834438ed3ebbc31 with parent 519d892f9523fe40cc11ec07323ffc2792614742, future [comparison URL](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (debc5abdd3fb55de560daf90c834438ed3ebbc31): [comparison URL](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6%  | [0.5%, 7.4%]    | 5     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -3.6% | [-13.3%, -0.5%] | 5     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.9%, -1.7%]  | 3     |
| All ❌✅ (primary)                 | -0.5% | [-13.3%, 7.4%]  | 10    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.4%, 1.3%] | 2     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.0%, 2.3%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.4%, 1.3%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=519d892f9523fe40cc11ec07323ffc2792614742&end=debc5abdd3fb55de560daf90c834438ed3ebbc31&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.3%]   | 6     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.5%, 0.3%]  | 7     |
</details>

**Bootstrap**: 672.189s -> 671.042s (-0.17%)
**Artifact size**: 315.72 MiB -> 315.75 MiB (0.01%)

Nilstrieb commented:
No regression, not that I expected any change.
@bors r+ rollup=maybe

bors commented:
:pushpin: Commit 083ee8369bb76bb130f6e60e0d0c939ee929c345 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb 083ee8369bb76bb130f6e60e0d0c939ee929c345 -->
<!-- homu: {"type":"Approved","sha":"083ee8369bb76bb130f6e60e0d0c939ee929c345","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- 
https://github.com/rust-lang/rust/pull/123143#issuecomment-2024087377

Philippe-Cholet committed:
Add invariant to VecDeque::pop_* that len < cap if pop successful

Similar to #114370 for VecDeque instead of Vec. It now uses `core::hint::assert_unchecked`.

Nilstrieb commented on tests/codegen/vecdeque_pop_push.rs:
```diff
@@ -0,0 +1,67 @@
+
+#[no_mangle]
+// CHECK-LABEL: @push_back_byte(
+pub fn push_back_byte(v: &mut VecDeque<u8>) {
```
thanks for including those tests btw, having the test to make sure the above tests are correct is super important!

Philippe-Cholet replied:
I merely mimicked the `Vec` version of this PR.
The 1st iteration was too much `Vec`-oriented.

Nilstrieb commented:
@bors r+

bors commented:
:pushpin: Commit 7a2678de7d3c0caa169542bd56a4fbf41b73a5b5 has been approved by `Nilstrieb`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nilstrieb 7a2678de7d3c0caa169542bd56a4fbf41b73a5b5 -->
<!-- homu: {"type":"Approved","sha":"7a2678de7d3c0caa169542bd56a4fbf41b73a5b5","approver":"Nilstrieb","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
